tests/test-doc-portal.c

```diff
@@ -687,6 +687,55 @@ test_add_named (void)
   assert_doc_not_exist (id1, basename1, "com.test.App2");
 }
 
+static void
+test_get_host_paths (void)
+{
+  g_autofree char *doc_id = NULL;
+  g_autofree char *expected_real_path = NULL;
+  g_autofree char *real_path = NULL;
+  const char *basename = "host-path";
+  g_autoptr(GVariant) path= NULL;
+  g_autoptr(GVariant) reply = NULL;
+  g_autoptr(GVariant) result = NULL;
+  g_autoptr(GVariantIter) iter = NULL;
+  GVariantBuilder builder;
+  GError *error = NULL;
+  const gchar* key = NULL;
+
+  if (!check_fuse_or_skip_test ())
+    return;
+
+  doc_id = export_new_file (basename, "content", FALSE);
+
+  g_variant_builder_init (&builder, G_VARIANT_TYPE ("(as)"));
+  g_variant_builder_open (&builder, G_VARIANT_TYPE ("as"));
+  g_variant_builder_add (&builder, "s", doc_id);
+  g_variant_builder_close (&builder);
+
+  reply = g_dbus_connection_call_sync (session_bus,
+                                       "org.freedesktop.portal.Documents",
+                                       "/org/freedesktop/portal/documents",
+                                       "org.freedesktop.portal.Documents",
+                                       "GetHostPaths", g_variant_builder_end (&builder),
+                                       G_VARIANT_TYPE ("(a{say})"),
+                                       0, -1,
+                                       NULL,
+                                       &error);
+
+  g_assert_no_error (error);
+  result = g_variant_get_child_value (reply, 0);
+
+  g_assert (g_variant_is_of_type (result, G_VARIANT_TYPE ("a{say}")));
+
+  expected_real_path = g_build_filename (outdir, basename, NULL);
+  iter = g_variant_iter_new (result);
+  while (g_variant_iter_loop (iter, "{&s@ay}", &key, &path)) {
```

Comment:
If this loop iters more than once, the key and path variables will leak

Reply:
I think `key` is not going to be leaked according to the docs:
```
For example, if you are only looping over simple integer and string types, g_variant_iter_next() is definitely preferred. For string types, use the ‘&’ prefix to avoid allocating any memory at all (and thereby avoiding the need to free anything as well).
```

Reply:
```
const gchar* path = NULL;
...
while (g_variant_iter_loop (iter, "{&s^&ay}", &key, &path))
  {
```
That should prevent leaking the value as well.

Reply:
Shouldn't it be `"{&s&@ay}"`?

Reply:
No, `@` gives you a GVariant:

> Used as a prefix for a GVariant type string (not a prefix for a format string, so as is a valid format string but @^as is not).

on the other hand:

> ^&ay 		equivalent to [g_variant_get_bytestring()](https://docs.gtk.org/glib/method.Variant.get_bytestring.html)

and from g_variant_get_bytestring:

> The data is owned by the instance.

Reply:
I understand now. Thank you for explaining.